text/0000-try-from.md

```diff
@@ -0,0 +1,165 @@
+}
+
+pub trait TryInto<T>: Sized {
+    type Err;
+
+    fn try_into(self) -> Result<T, Self::Err>;
+}
+```
+
+In a fashion similar to `From` and `Into`, a blanket implementation of `TryInto`
+is provided for all `TryFrom` implementations:
+
+```rust
+impl<T, U> TryInto<U> for T where U: TryFrom<T> {
+    type Error = U::Err;
+
+    fn try_into(self) -> Result<U, Self::Err> {
+        U::try_from(self)
+    }
+}
+```
+
+In addition, implementations of `TryFrom` will be provided to convert between
+*all combinations* of integer types:
+
+```rust
+#[derive(Debug)]
+pub struct TryFromIntError(());
+
+impl fmt::Display for TryFromIntError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str(self.description())
+    }
+}
+
+impl Error for TryFromIntError {
+    fn description(&self) -> &str {
+        "out of range integral type conversion attempted"
+    }
+}
+
+impl TryFrom<usize> for u8 {
+    type Err = TryFromIntError;
+
+    fn try_from(t: usize) -> Result<u8, TryFromIntError> {
+        // ...
+    }
+}
+
+// ...
+```
+
+This notably includes implementations that are actually infallible, including
+implementations between a type and itself. A common use case for these kinds
+of conversions is when interacting with a C API and converting, for example,
+from a `u64` to a `libc::c_long`. `c_long` may be `u32` on some platforms but
+`u64` on others, so having an `impl TryFrom<u64> for u64` ensures that
+conversions using these traits will compile on all architectures.  Similarly, a
+conversion from `usize` to `u32` may or may not be fallible depending on the
+target architecture.
+
+The standard library provides a reflexive implementation of the `From` trait
+for all types: `impl<T> From<T> for T`. We could similarly provide a "lifting"
+implementation of `TryFrom`:
+
+```rust
+impl<T, U: From<T>> TryFrom<T> for U {
+    type Err = Void;
+
+    fn try_from(t: T) -> Result<U, Void> {
+        Ok(U::from(t))
+    }
+}
+```
+
+However, this implementation would directly conflict with our goal of having
+uniform `TryFrom` implementations between all combinations of integer types. In
+addition, it's not clear what value such an implementation would actually
+provide, so this RFC does *not* propose its addition.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It is unclear if existing fallible conversion traits can backwards-compatibly
+be subsumed into `TryFrom` and `TryInto`, which may result in an awkward mix of
+ad-hoc traits in addition to `TryFrom` and `TryInto`.
+
+# Alternatives
+[alternatives]: #alternatives
+
+We could avoid general traits and continue making distinct conversion traits for
+each use case.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Are `TryFrom` and `TryInto` the right names? There is some precedent for the
+`try_` prefix: `TcpStream::try_clone`, `Mutex::try_lock`, etc.
+
+What should be done about `FromStr`, `ToSocketAddrs`, and other ad-hoc fallible
```

Comment:
>It's also worth noting that the fate of parse is basically unrelated to the fate of FromStr.

`parse` itself is not affected, it'll just get a generic duplicate `try_into` if `TryFrom<str>` is implemented for the same types as `FromStr` and `TryInto` gets the blanket impl based on `TryFrom`.
And I'm saying that this duplicate will almost never be used, because `parse` is preferable in concrete code and probably never used in code generic enough to not be able to use `parse`.

>Why do you feel that impl FromStr for Foo is an intrinsically more meaningful implementation than impl<'a str> TryFrom<&'a str> for Foo?

I don't feel that way with respect to `FromStr` in particular, `FromStr` is already meaningless enough :)